[Diag] Create a serialized format for Localized Diagnostics

HassanElDesouky opened the pull request:
<!-- What's in this pull request? -->
Introduces a serialized file format and create a tool for converting the diagnostic file from `.yaml` to `.db`

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

HassanElDesouky commented:
cc @xedin

xedin commented on include/swift/AST/LocalizationFormat.h:
```diff
@@ -28,6 +30,92 @@ namespace swift {
+  llvm::raw_fd_ostream OS;
+
+public:
+  SerializedLocalizationWriter(llvm::StringRef filePath)
```
Let's re-design interface here a little bit - make it:

1. `SerializedLocalizationWriter` should only have `generator` as a field;
2. Constructor could be removed - it's incorrect to mutate file before any data is added to the hash table.

3. Let's separate logic from destructor into a separate `emit` method.
```
void insert(llvm::StringRef id, llvm::StringRef translation);
bool emit(llvm::StringRef filePath) const;
```

4. Destructor could be removed.

xedin commented on include/swift/AST/LocalizationFormat.h:
```diff
@@ -43,7 +131,29 @@ class LocalizationProducer {
+class SerializedLocalizationProducer final : public LocalizationProducer {
+  using SerializedLocalizationTable =
+      llvm::OnDiskIterableChainedHashTable<LocalizationLookupInfo>;
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> Buffer;
```
We need to make sure that it's not an error but indeed a `MemoryBuffer` because if it's an error it has to be reported and localization couldn't be used. So the field should be `std::unique_ptr<llvm::MemoryBuffer>`.

xedin commented on include/swift/AST/LocalizationFormat.h:
```diff
@@ -43,7 +131,29 @@ class LocalizationProducer {
+  using SerializedLocalizationTable =
+      llvm::OnDiskIterableChainedHashTable<LocalizationLookupInfo>;
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> Buffer;
+  llvm::BitstreamCursor Cursor;
```
Why are we storing cursor?

xedin reviewed (changes requested):
Direction looks good! I have left some comments in line.

xedin commented on include/swift/AST/LocalizationFormat.h:
```diff
@@ -43,7 +131,29 @@ class LocalizationProducer {
+        Cursor(*Buffer.get()) {
+    auto base = Cursor.getBitcodeBytes();
+    const uint8_t *B = base.data();
+    uint32_t offset = endian::readNext<uint32_t, little, unaligned>(B);
```
I think instead of even using `Cursor` here, you could use `getBufferStart()` because these "offset" bytes are always stored at `0`, so `Cursor` could be removed completely.

xedin commented on include/swift/AST/LocalizationFormat.h:
```diff
@@ -43,7 +131,29 @@ class LocalizationProducer {
+
+public:
+  explicit SerializedLocalizationProducer(llvm::StringRef filePath)
+      : Buffer(llvm::MemoryBuffer::getFileOrSTDIN(filePath)),
```
Since we need to verify that `getFileOrSTDIN` returns a valid buffer, I'd suggest we pass `std::unique_ptr<MemoryBuffer>` as a constructor argument so caller handle opening of the buffer and producing a warning if needed.

HassanElDesouky commented on include/swift/AST/DiagnosticEngine.h:
```diff
@@ -24,6 +24,7 @@
 #include "swift/AST/TypeLoc.h"
 #include "llvm/ADT/StringSet.h"
 #include "llvm/Support/Allocator.h"
+#include "llvm/Support/CommandLine.h"
```
Thinking about it, I definitely don't need `CommandLine.h`. However, I need `Path.h` that's mainly because `llvm::sys` namespace is available in it. ([Check out the llvm docs](https://llvm.org/doxygen/Path_8h.html)) 

I'm using `llvm::sys::path` in `setLocalization` 
https://github.com/apple/swift/pull/33022/files#diff-3c1e580ca884168dd6e95ebf168c9db3R747
WDYT?

xedin replied:
Sure! In general it should include everything it uses.

HassanElDesouky commented on include/swift/AST/DiagnosticEngine.h:
```diff
@@ -750,13 +750,17 @@ namespace swift {
+            std::make_unique<diag::SerializedLocalizationProducer>(file.get());
       } else {
         llvm::sys::path::replace_extension(filePath, ".yaml");
-        assert(llvm::sys::fs::exists(filePath));
```
I was using assertion, but this will fail with other languages that don't have translation files yet. So, I think we should fallback to default messages until we remove messages from `.def` files and then we can fallback to messages from the English file i.e. (`en.db`)

HassanElDesouky commented on lib/AST/LocalizationFormat.cpp:
```diff
@@ -76,17 +73,38 @@ template <> struct MappingTraits<DiagnosticNode> {
+    offset = endian::readNext<uint32_t, little, unaligned>(base);
+  }
+  auto base =
+      reinterpret_cast<const uint8_t *>(Buffer.get()->getBuffer().data());
```
Is the use of `reinterpret_cast` correct here?

HassanElDesouky replied:
and if it's... should I use `reinterpret_cast` in `ReadKey` and `ReadData` here: https://github.com/apple/swift/pull/33022/files#diff-d53379d6eb36fd21470769bbf076697cR95-R102

HassanElDesouky commented on lib/AST/LocalizationFormat.cpp:
```diff
@@ -76,17 +73,38 @@ template <> struct MappingTraits<DiagnosticNode> {
 }
 
+SerializedLocalizationProducer::SerializedLocalizationProducer(
+    std::unique_ptr<llvm::MemoryBuffer> &buffer)
```
Since I'm passing a `unique_ptr`. I can't pass it by value (make a copy). Therefore, I passed it by reference. Does that make sense?

xedin commented on lib/AST/LocalizationFormat.cpp:
```diff
@@ -64,15 +73,53 @@ template <> struct MappingTraits<DiagnosticNode> {
+
+llvm::StringRef SerializedLocalizationProducer::getMessageOr(
+    swift::DiagID id, llvm::StringRef defaultMessage) const {
+  if (!SerializedTable) {
```
Can this ever be true?

xedin commented on tools/swift-serialize-diagnostics/swift-serialize-diagnostics.cpp:
```diff
@@ -0,0 +1,80 @@
+static llvm::cl::OptionCategory Category("swift-serialize-diagnostics Options");
+
+static llvm::cl::opt<std::string>
+    InputFilePath("input-file-path", llvm::cl::desc("Path of the input file"),
```
`of the` -> `to the` and probably worth to mention `YAML`.

xedin commented on tools/swift-serialize-diagnostics/swift-serialize-diagnostics.cpp:
```diff
@@ -0,0 +1,80 @@
+    return 1;
+  }
+
+  YAMLLocalizationProducer YAMLDiagnostics(options::InputFilePath);
```
Let's just call it `yaml`.

xedin commented on tools/swift-serialize-diagnostics/swift-serialize-diagnostics.cpp:
```diff
@@ -0,0 +1,80 @@
+  llvm::sys::path::replace_extension(SerializedFilePath, ".db");
+
+  SerializedLocalizationWriter Serializer;
+  for (unsigned id = 0; id < YAMLDiagnostics.diagnostics.size(); ++id)
```
I'd suggest we make `YAMLLocalizationProducer` to conform to iterator interface so we could simply use:

```c++
for (const auto &translation : yaml) {
  Serializer.insert(diagnosticID[translation.first], translation.second);
}
```

xedin commented on include/swift/AST/LocalizationFormat.h:
```diff
@@ -28,6 +31,85 @@ namespace swift {
+
+public:
+  void insert(llvm::StringRef id, llvm::StringRef translation);
+  void emit(llvm::StringRef filePath);
```
Let's make `emit` return `bool` to indicate whether translations have been successfully emitted into the file.

xedin commented on tools/swift-serialize-diagnostics/swift-serialize-diagnostics.cpp:
```diff
@@ -0,0 +1,80 @@
+  for (unsigned id = 0; id < YAMLDiagnostics.diagnostics.size(); ++id)
+    Serializer.insert(diagnosticID[id], YAMLDiagnostics.diagnostics[id]);
+
+  Serializer.emit(SerializedFilePath.str());
```
Let's check whether `emit` has successful and if not - write an error out and return `-1;`

xedin commented on localization/diagnostics/en.yaml:
```diff
@@ -7122,31 +7122,6 @@
     %0 is not a 'func', 'init', 'subscript', or 'var' computed property
     declaration
 
-- id: autodiff_attr_accessor_not_found
```
It's okay to commit this is a separate PR btw.

xedin reviewed (changes requested):
Getting closer! I've left some more comments inline.

HassanElDesouky commented on include/swift/AST/DiagnosticEngine.h:
```diff
@@ -742,8 +743,25 @@ namespace swift {
+      if (llvm::sys::fs::exists(filePath)) {
+        auto file = llvm::MemoryBuffer::getFile(filePath);
+        localization =
+            std::make_unique<diag::SerializedLocalizationProducer>(file.get());
```
I thought that since we are _sure_ that the file exists (because we already check at L752 `if (llvm::sys::fs::exists(filePath)) `. We don't need to verify that we have a file buffer, because we will always do have a file buffer. WDYT?

xedin replied:
Existence is one thing, but there could be I/O error while reading or memory mapping it.

HassanElDesouky committed:
[Locale] Serialize YAML to an OnDiskHashTable format and create a tool for serialization

xedin committed:
[Localization] Simplify implementation of iterator over YAML

xedin committed:
[Localization] Serialization writer should own the inserted data

xedin committed:
[Localization] Adjust types associated with reader/writer info

Add more usages of `offset_type` where appropriate and change
buffer type to `const unsigned char *`.

xedin commented:
@swift-ci please build toolchain

xedin committed:
[Localization] Adjust cmake to include serialized diagnostics in toolchain

Building of the diagnostic database to be done in three steps:

- Copy YAMLs from <source>/localization/diagnostics to <build>/share/swift/diagnostics
- Serialize English into <build>/share/swift/diagnostics
- Install everything (YAML and `.db`) from <build>/share/swift/diagnostics
  to share/swift/diagnostics in toolchain

xedin commented:
@swift-ci please build toolchain

owenv commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please build toolchain macOS platform

HassanElDesouky commented:
@xedin why did the build failed? https://ci.swift.org/job/swift-PR-toolchain-osx/lastBuild/console

xedin commented:
CI issue, let's try again.

xedin commented:
@swift-ci please build toolchain macOS platform

xedin commented:
@swift-ci please build toolchain macOS platform

xedin commented:
@shahmishal Looks like there is a problem with github auth on one of the machines.

shahmishal commented:
Should be fixed, please try again

xedin commented:
@swift-ci please build toolchain macOS platform

xedin commented:
@shahmishal Thanks, looks like it worked!

xedin commented:
Great! Looks like we got `.db` in the toolchain now, I can make it so everything is placed into the `share/swift/diagnostics` and copied from there

xedin commented:
@swift-ci please clean build toolchain

xedin commented:
@swift-ci please build toolchain

xedin committed:
[Localization] NFC: Include all of the headers based on the types used

xedin committed:
[Localization] Switch serialization format to use DiagID as a key

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please asan test

xedin commented:
@swift-ci please ASAN test

xedin commented:
@shahmishal Both toolchain builds were successful but none updated their state :/

xedin commented:
Ok, both toolchains look good now. Let's wait for the tests to run and we'll be good to go with this.

xedin commented:
ASAN failure is in GSB and is unrelated to these changes.

xedin committed:
[Localization] Make return type of the serialization main cross platform
